### PR TITLE
Fix issue cloning BLangLiteral

### LINKED ISSUE #36172: Compiler crashes with NPE for the valid code
**Description:**
$subject.

```bal
type ReadableTextRecordChannel any;

public const COLON = ":";

public const TAB = "\t";

public const COMMA = ",";

public type Separator COMMA|TAB|COLON|string;

public class ReadableCSVChannel {
    private ReadableTextRecordChannel? dc;

    isolated function init(Separator fs = ",") {
    }
}
```

<details>
  <summary><b>stacktrace</b>  (<i>click to expand</i>)</summary>

```
2022-05-17 13:58:19,352] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:405)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:363)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1070)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:235)
	at org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable.accept(BLangSimpleVariable.java:59)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4155)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4131)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:4112)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:467)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:235)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:81)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4155)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4131)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.addClassMemberFunctionsToTopLevel(Desugar.java:506)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.addAttachedFunctionsToPackageLevel(Desugar.java:489)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.visit(Desugar.java:752)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:167)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.rewrite(Desugar.java:8434)
	at org.wso2.ballerinalang.compiler.desugar.Desugar.perform(Desugar.java:456)
	at io.ballerina.projects.internal.CompilerPhaseRunner.desugar(CompilerPhaseRunner.java:216)
	at io.ballerina.projects.internal.CompilerPhaseRunner.performBirGenPhases(CompilerPhaseRunner.java:152)
```
</details>

**Affected Versions:**
current master

### PR DESCRIPTION
**Pull request by lochana-chathura**
## Purpose
$subject.

Fixes #36172 

## Approach
N/A

## Samples
N/A

## Remarks
N/A

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java [hunk 1]
```diff
@@ -348,7 +348,6 @@ private void cloneBLangLiteral(BLangLiteral source, BLangLiteral clone) {
 
         clone.value = source.value;
         clone.originalValue = source.originalValue;
-        clone.isFiniteContext = source.isFiniteContext;
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java:351]: Clone or original I believe type checking should follow the same and algorithms should not be dependent on order if we change the order between runs. 

1. Shouldn't we propagate finalized values back to cloned nodes as well?

2. Removing this line does not stop cloning but now we have a default value `false` only for this type that we may be fooled to consider in other places as a cloned value.

Comment (lochana-chathura) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java:351]: One can refer to https://github.com/ballerina-platform/ballerina-lang/issues/36172#issuecomment-1129216294 for details on this change.

Comment (lochana-chathura) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java:351]: Change we agreed in our offline discussion added with [2826b05](https://github.com/ballerina-platform/ballerina-lang/pull/36177/commits/2826b0548e610dbe618158a296ae01b4ff36f2e0).

**Review by KRVPerera — APPROVED** [review]

**Comment by lochana-chathura** [comment]
I have tested and verified the `ballerina-io` build on top of this fix.

**Comment by pcnfernando** [comment]
Shouldn't this go to 2201.1.0-stage as well even though the build failure is on master @lochana-chathura?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -713,6 +713,7 @@ private BType getTypeOfHexFloatingPointLiteral(BLangLiteral literalExpr, Object
     }
 
     public BType setLiteralValueAndGetType(BLangLiteral literalExpr, BType expType, AnalyzerData data) {
+        literalExpr.isFiniteContext = false;
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:716]: @KRVPerera can you remember the reason behind this verdict

Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:716]: This is by default false. Logic inside this function is the one which set this to true. Original problem is some nodes are getting cloned at desugar had this set to true. If this is set to true it prevents  type checker progressing and setting the result.

**Review by pcnfernando — APPROVED** [review]

**Pull request merged**